full_storage_utilization_test: reclaim and scale out tests

cezarmoise opened the pull request:
Adds test cases and jenkins files for 90% storage utilization tests.
Covers these test cases:

- Space reclaim
- [x] https://github.com/scylladb/scylla-cluster-tests/issues/9169 
- [x] https://github.com/scylladb/scylla-cluster-tests/issues/9166 
- [x] https://github.com/scylladb/scylla-cluster-tests/issues/9167
- Scale out
- [x] https://github.com/scylladb/scylla-cluster-tests/issues/9256
- [x] https://github.com/scylladb/scylla-cluster-tests/issues/9257  
  These test cases are implemented as performance tests, to check that scaling out and scaling in do not cause performance issue at 90% utilization, under a 50% load.
- [x] https://github.com/scylladb/scylla-cluster-tests/issues/9597

The tests are meant to be run manually and checking their monitoring stack is part of their passing condition.

refs: https://github.com/scylladb/scylla-cluster-tests/issues/9129

### [Argus Dashboard of all cases](https://argus.scylladb.com/test_runs?state=WyJkMzE4ZWZjMi05MWIxLTQ5Y2UtOWM0Yi1mNmMwZjBhMzkyMDAiLCJiNDc0NjEzOC02OGI5LTQ0NGQtOGFlNi1jZTA3ZTZkYmMzMGUiLCI4Y2I5NWZkMS01MjdkLTQ5OTMtYjhmOC0wZjYyMTZmOTczZGUiLCIyOWJlYTNkZC0wNDljLTRhNjQtYjM2OC1kYTA2NTk2OGZlYzQiLCI4M2YzYWZmZi0yNTcyLTQ4MjYtYTJiMC0xNWM5ZTE0ZWFjZDgiLCI5MGU1ZDgwYi04Y2NlLTRkOGMtODgxNy1jM2FlYWYyNDE4MTMiXQ)

cezarmoise commented:
https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/1b0e85b6817bd400a91761886c3cb1853a944ca3

Also include test keyspaces when reconfiguring replication strategy to include the new dc.

Last test run: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/cezar/job/byo-longevity-test/69/consoleFull 

```
04:34:09  < t:2024-11-21 02:34:06,572 f:full_storage_utilization_test_2.py l:131  c:FullStorageUtilizationTest2 p:INFO  > Node     Total GB     Used GB      Avail GB     Used %  
04:34:29  < t:2024-11-21 02:34:28,756 f:full_storage_utilization_test_2.py l:143  c:FullStorageUtilizationTest2 p:INFO  > 1        436          396          40           91.0%
04:34:51  < t:2024-11-21 02:34:50,943 f:full_storage_utilization_test_2.py l:143  c:FullStorageUtilizationTest2 p:INFO  > 2        436          393          44           90.0%
04:35:13  < t:2024-11-21 02:35:13,134 f:full_storage_utilization_test_2.py l:143  c:FullStorageUtilizationTest2 p:INFO  > 3        436          395          42           91.0%
04:35:36  < t:2024-11-21 02:35:35,329 f:full_storage_utilization_test_2.py l:143  c:FullStorageUtilizationTest2 p:INFO  > 4        436          403          34           93.0%
04:35:36  < t:2024-11-21 02:35:35,851 f:full_storage_utilization_test_2.py l:143  c:FullStorageUtilizationTest2 p:INFO  > 5        436          37           400          9.0%
04:35:58  < t:2024-11-21 02:35:58,183 f:full_storage_utilization_test_2.py l:143  c:FullStorageUtilizationTest2 p:INFO  > 6        436          37           400          9.0%
04:36:21  < t:2024-11-21 02:36:20,536 f:full_storage_utilization_test_2.py l:143  c:FullStorageUtilizationTest2 p:INFO  > 7        436          37           400          9.0%
04:36:21  < t:2024-11-21 02:36:20,536 f:full_storage_utilization_test_2.py l:153  c:FullStorageUtilizationTest2 p:INFO  > Cluster  3052         1698         1360         56.0%
```

Did not redistribuite data to new dc.

cezarmoise commented:
https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/b2e8d565f4a47de0c0442375006cba40ecb7efba

Added code to put usage at certain points in the results tab, e.g. 
![Screenshot 2024-11-21 230618](https://github.com/user-attachments/assets/00a446fa-0c7b-4391-987c-0eb17d0f1e12)

pehala commented:
How long do think, before this PR is reviewable and out of draft?

pehala commented on sdcm/nemesis.py:
```diff
@@ -5640,6 +5677,24 @@ def disrupt(self):
+        self.grow_cluster()
+
+
+class GrowFillMoneky(Nemesis):
```
Will these be run also on a normal longevities?

cezarmoise replied:
No, I named the functions without the `distrupt_` prefix. That's why I named it `Monkey` instead of `Nemesis`.
The only difference is that it is not shown in the Argus tab for nemesis, but it doesn't matter since we run it alone.
Technically, `GrowShrinkClusterNemesis` could be used instead of adding a new one.

pehala commented on sdcm/nemesis.py:
```diff
@@ -5640,6 +5677,24 @@ def disrupt(self):
+        self.grow_cluster()
+
+
+class GrowFillMoneky(Nemesis):
```
s/Moneky/Monkey

pehala commented on test-cases/features/elasticity/elasticity-90-percent-i4i-2xlarge-drop.yaml:
```diff
@@ -0,0 +1,31 @@
+nemesis_interval: 30
+
+user_prefix: 'elasticity-test'
+ami_id_db_scylla_desc: 'VERSION_DESC'
```
Why do we need this?

pehala commented on data_dir/nemesis.yml:
```diff
@@ -257,7 +268,6 @@
 - disrupt_rolling_restart_cluster:
   - disruptive = True
   - kubernetes = True
-  - config_changes = True
```
was trhis change done automatically?

cezarmoise replied:
Yes, running the unit test locally changes some files.

pehala commented on jenkins-pipelines/oss/features/elasticity/elasticity-90-percent-truncate.jenkinsfile:
```diff
@@ -0,0 +1,19 @@
+    test_config: """["test-cases/features/elasticity/elasticity-90-percent-i4i-2xlarge-truncate.yaml", "configurations/disable_kms.yaml"]""",
+    scylla_version: 'masters:latest',
+    email_recipients: 'qa@scylladb.com',
+    post_behavior_db_nodes: 'destroy',
```
I do not think we need to define as many variables as we have here

pehala commented on performance_regression_test.py:
```diff
@@ -613,21 +613,21 @@ def test_latency(self):
         self.run_fstrim_on_all_db_nodes()
         self.preload_data()
-        self.wait_no_compactions_running(n=160)
+        self.wait_no_compactions_running(n=240)
```
Why? And what will its effect be on rest of the performance tests

cezarmoise replied:
After the XFS fix we started writing more data to clusters, and for really big instances it didn't finish compactions in time.
`n` in this case is the number of retries, and it retries once per minute.
It won't effect other tests except by preventing timeout failures.

pehala commented on sdcm/nemesis.py:
```diff
@@ -2068,8 +2068,8 @@ def _prepare_test_table(self, ks='keyspace1', table=None):
         if ks not in test_keyspaces or not table_exist:
-            stress_cmd = "cassandra-stress write n=400000 cl=QUORUM -mode native cql3 " \
-                         f"-schema 'replication(strategy=NetworkTopologyStrategy," \
+            stress_cmd = f"cassandra-stress write n=400000 cl=QUORUM -mode native cql3 " \
```
WIll change affect other nemesis as well?

cezarmoise replied:
Yes, this is a fix. Because the keyspace name was not added to the schema, it would never be created with the name we wanted.
I noticed when running `disrupt_drop`. It didn't recreate the table, so drop failed. Truncate was not affected since it leaves the table behind.

cezarmoise commented:
Pushed https://github.com/scylladb/scylla-cluster-tests/commit/332d00d312135677fb3bca2e8f8c152a321936ed with PR fixes:
- rebased to master
- fixed Monkey typo
- removed extra variables in jenkinfsfiles that were just the defaults

pehala commented:
Changing to draft, as this PR needs to be adjusted to use 3 racks

cezarmoise commented on performance_regression_test.py:
```diff
@@ -630,21 +634,21 @@ def test_latency(self):
         self.run_fstrim_on_all_db_nodes()
         self.preload_data()
-        self.wait_no_compactions_running(n=160)
+        self.wait_no_compactions_running(n=240)
```
This needed increased because is was too short for tests that fill an 8XL cluster to 90%

cezarmoise commented on data_dir/nemesis.yml:
file auto-updated by unit tests

cezarmoise commented on data_dir/nemesis_classes.yml:
file auto-updated by unit tests

cezarmoise commented on performance_regression_test.py:
```diff
@@ -219,6 +220,9 @@ def _stop_load_when_nemesis_threads_end(self):
     @optional_stage('perf_preload_data')
     def preload_data(self, compaction_strategy=None):
+        # Check if the test keyspace should be pre-created
+        if self.params.get('pre_create_keyspace'):
```
This allows pre-creating the test keyspaces and set initial tablets to them, same as longevity tests.

cezarmoise commented on sdcm/nemesis.py:
```diff
@@ -4240,6 +4250,14 @@ def add_new_nodes(self, count, rack=None, instance_type: str = None) -> list[Bas
         return nodes
 
+    @latency_calculator_decorator(legend="Adding new nodes to racks")
+    def add_new_nodes_to_racks(self, count, racks, instance_type: str = None) -> list[BaseNode]:
```
Compared to `add_new_nodes`, this allows adding nodes to different racks without waiting for tablet migration

cezarmoise commented:
Changes:
- b516630f937aef6b130bff5d42424109a7e2732f
  - cherry-picked from https://github.com/scylladb/scylla-cluster-tests/pull/10099 to allow setting initial tablets

- ef341e26be3687e412321d48c79c68424c241ebe
  - add new methods for growing a cluster on each rack
  - this could probably be merged with the existing methods

- e3253a6e14c24e6980100c000b3dee1ccafc4104
  - updated all tests to use 3 availability zones (3 racks)
  - updated all tests to use 3 loaders (1 per rack)
  - updated all tests' initial write commands to have 1 stress write per loader
  - improve amount of data in the initial write to be closer to 90%

cezarmoise commented on sdcm/tester.py:
```diff
@@ -3342,6 +3342,7 @@ def check_latency_during_ops(self, hdr_tags: list[str]):
                                                   node_benchmarks=benchmarks_results,
                                                   email_subject_postfix=self.params.get('email_subject_postfix'))
             except Exception as exc:  # noqa: BLE001
+                self.log.exception("Error while checking regression")
```
This logs a traceback of the error.

For example, if a test does not have `use_hdrhistogram: true`, then the error would be 
> 2025-03-02 13:55:20.759: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=f0a49a5c-6891-43c3-8f99-85b5798e090f, source=PerformanceRegressionTest.check_regression() message=Failed to check regression
exception='NoneType' object is not iterable

which is not useful at all

cezarmoise commented:
https://github.com/scylladb/scylla-cluster-tests/commit/3a5d3fb88c3f426ba4a80e030fa7e788484782bb

- cleaned up commits
- removed commits that were unrepeated (moved to other PRs)
- use GrowShrinkClusterNemesis instead of writing new code

pehala commented on data_dir/nemesis.yml:
```diff
@@ -18,20 +18,24 @@
   - run_with_gemini = False
   - limited = True
   - topology_changes = True
+- disrupt_add_remove_dc: []
```
I would say this nemesis should have some flags

pehala replied:
I know, but it means you did not include any flags for your nemesis. I think it should be marked appropriately

cezarmoise replied:
`disrupt_add_remove_dc` is an existing nemesis, not added by this PR

Lakshmipathi commented on test-cases/features/elasticity/elasticity-90-percent-perf-i4i-large-grow-i4i-4xlarge.yaml:
```diff
@@ -0,0 +1,43 @@
+use_hdrhistogram: true
+email_recipients: ["scylla-perf-results@scylladb.com"]
+email_subject_postfix: 'elasticity test'
+parallel_node_operations: true
```
we dont need `auto_snapshot: false` for this config file?

cezarmoise replied:
I only used that for the reclaim tests, so as to make it faster by not waiting for compaction to reclaim space.

cezarmoise commented:
https://github.com/scylladb/scylla-cluster-tests/commit/739c4f4e7820826fc8cd246d3ea194387b67373b: rebase to fix PR checks

cezarmoise commented:
https://github.com/scylladb/scylla-cluster-tests/commit/295827a8ba25756a97f717e7d694be17bd2a1210 use pre_create_keyspace for ttl case instead of user profile.

cezarmoise commented:
https://github.com/scylladb/scylla-cluster-tests/commit/734bfb4cf03f78bee3953d3ae55b90d4bf9d3d41 update workload on grow larger test case to be the same as grow smaller, since they use the same instance types

cezarmoise commented:
Ci is failing due to https://github.com/scylladb/scylla-cluster-tests/issues/10450

roydahan reviewed (changes requested):
While I understand the approach of trying to utilize the peformance regression test with nemesis, the new nemesis that were added here aren't comply with the concept of nemesis and they will probably break all other longevities.

If you do it only to utlize the context manager of measuring latency or something like that, there are are probably better ways to do it.

Also, the commit message isn't descriptive enough to explain the motivations and the changes you want to introduce and their reasons.

roydahan commented:
One more thing, regardless of my review above.
If you would want to introduce new nemesis (which again, shouldn't be the case IMO here), the testing of this PR should have cover much much more than your cases, proving they won't break our longeivites.
Sample of such testing:
1. Several longevities running only these nemesis multiple times.
2. Several longevities running multiple nemesis which this nemesis is part of them.
3. Parallel longevities that includes this nemesis from one of our existing cases.

vponomaryov commented on sdcm/nemesis.py:
```diff
@@ -4336,6 +4346,29 @@ def _double_cluster_load(self, duration: int) -> None:
+        self._grow_cluster(rack=None)
+        time.sleep(sleep_time_between_ops)
+
+        # write more data to the cluster
+        stress_cmds = self.cluster.params.get('stress_cmd')
+        stress_queue = []
+        for stress_cmd in stress_cmds:
+            stress_queue.append(self.tester.run_stress_thread(stress_cmd=stress_cmd,
+                                stop_test_on_failure=False, stats_aggregate_cmds=False, round_robin=True))
+        for stress in stress_queue:
+            self.tester.verify_stress_thread(stress)
```
Agree with @roydahan 

About the current approach:
At first, many scenarios have only "reads" as a main stress configured in the `stress_cmd` config option, reads do not add more data, obviously.
At second, above comment `write more data to the cluster` is false.
Running the same command we create load, but we re-write the same data.
So, there will be no `more data` and no change in the used disk space.

cezarmoise replied:
My understanding was that if the function is not named `distrupt_`, it will not be picked up by Sisyphus.
But I refactored it out in https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/605e613008c6ae56a02e433066aa155363c6ad1b

vponomaryov commented on sdcm/nemesis.py:
```diff
@@ -5791,6 +5824,15 @@ def disrupt(self):
         self.disrupt_add_remove_dc()
 
 
+class GrowFillMonkey(Nemesis):
```
+1

cezarmoise replied:
Removed, refactored as test_custom_time.

https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/d9f61e10b4846c8ddea0770a570e9d84d2d3c6ed and https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/605e613008c6ae56a02e433066aa155363c6ad1b

cezarmoise commented on sdcm/nemesis.py:
```diff
@@ -6035,6 +6077,16 @@ def disrupt(self):
         self.disrupt_nodetool_cleanup()
 
 
+class DropMonkey(Nemesis):
```
I didn't find any nemesis that just does a drop.
But I don't really need it since I can just put `post_prepare_cql_cmds`

Removed in https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/cf3c25725d7ed467c9e8ca0c5911d6c5b2632efd

cezarmoise commented:
Changes:
- https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/cf3c25725d7ed467c9e8ca0c5911d6c5b2632efd: uses only existing nemesis and parameters for tests
- https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/79e27826691139145e3dd1e475897e68e5b97340: updated commit message
- https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/d9f61e10b4846c8ddea0770a570e9d84d2d3c6ed: add parameter to allow longevity test to grow with a different cluster type than the initial cluster
- https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/605e613008c6ae56a02e433066aa155363c6ad1b: refactored old nemesis into longevity test

cezarmoise commented:
@scylladb/qa-maintainers please review

vponomaryov commented on test-cases/features/elasticity/elasticity-90-percent-i4i-2xlarge-truncate.yaml:
```diff
@@ -0,0 +1,42 @@
+prepare_write_cmd: [
+  "cassandra-stress write no-warmup cl=QUORUM n=480000000 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=200 -col 'size=FIXED(128) n=FIXED(8)' -pop seq=1..480000000",
+  "cassandra-stress write no-warmup cl=QUORUM n=480000000 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=200 -col 'size=FIXED(128) n=FIXED(8)' -pop seq=480000001..960000000",
+  "cassandra-stress write no-warmup cl=QUORUM n=480000000 -schema keyspace=ks_truncate 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=200 -col 'size=FIXED(128) n=FIXED(8)' -pop seq=960000001..1440000000",
```
We have nemesis `disrupt_truncate` which uses exactly the same keyspace name -> `ks_truncate`.
And the `nemesis_during_prepare` SCT config option is enabled by default...

So, to avoid possible unexpected problems, I recommend renaming all the new keyspaces here to some more unique names like `ks_90percent_util_truncate`...

cezarmoise replied:
renamed in https://github.com/scylladb/scylla-cluster-tests/pull/9305/commits/8e62b9fa81650c190269b5b98dca562883683fe1

vponomaryov commented on test-cases/features/elasticity/elasticity-90-percent-perf-i4i-2xlarge-grow-fill-i4i-large.yaml:
```diff
@@ -0,0 +1,45 @@
+test_duration: 2880
```
Where is this new config file is used?
Missing jenkinsfile for it?

pehala commented:
> @scylladb/qa-maintainers please review

re-ping

soyacz reviewed (commented):
It's very hard to verify if this PR is ok:
1. I miss docstrings for these cases - I propose to add intention on top of each yaml
2. e.g. `elasticity-90-percent-perf-i4i-8xlarge-grow-fill-i4i-large.yaml` test linked(??) to https://github.com/scylladb/scylla-cluster-tests/issues/9597#issuecomment-2633291581doesn't match - I can see drops of throughput, and actually I don't see throttling.
3. It's hard to look at selected test results... - I don't know which yaml refers to which run id.

Next time I propose to create separate PR's for each case.

cezarmoise commented:
closed if favor of https://github.com/scylladb/scylla-cluster-tests/pull/10999, https://github.com/scylladb/scylla-cluster-tests/pull/11000 and https://github.com/scylladb/scylla-cluster-tests/pull/11001
